File: src/Components/Builders/SigningContentBuilder.cs
using System.Text;
using HmacManager.Headers;

namespace HmacManager.Components;

/// <summary>
/// Provides methods to build and sign content for a request using specific parameters.
/// </summary>
public class SigningContentBuilder : ISigningContentBuilder
{
    /// <summary>
    /// Gets the <see cref="StringBuilder"/> used to construct the signing content.
    /// </summary>
    protected readonly StringBuilder Builder;

    /// <summary>
    /// Gets the context containing request-related information used in building the signing content.
    /// </summary>
    protected readonly SigningContentContext Context;

    /// <summary>
    /// Initializes a new instance of the <see cref="SigningContentBuilder"/> class.
    /// </summary>
    internal SigningContentBuilder()
    {
        Context = new();
        Builder = new StringBuilder();
    }

    /// <summary>
    /// Sets the HTTP request to be used in the signing content.
    /// </summary>
    /// <param name="request">The HTTP request message.</param>
    /// <returns>The <see cref="SigningContentBuilder"/> instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown when <paramref name="request"/> is <c>null</c>.</exception>
    internal SigningContentBuilder WithRequest(HttpRequestMessage request)
    {
        ArgumentNullException.ThrowIfNull(request, nameof(request));

        Context.Request = request;
        return this;
    }

    /// <summary>
    /// Sets the public key to be used in the signing content.
    /// </summary>
    /// <param name="publicKey">The public key as a <see cref="Guid"/>.</param>
    /// <returns>The <see cref="SigningContentBuilder"/> instance.</returns>
    internal SigningContentBuilder WithPublicKey(Guid publicKey)
    {
        Context.PublicKey = publicKey;
        return this;
    }

    /// <summary>
    /// Sets the date requested to be used in the signing content.
    /// </summary>
    /// <param name="dateRequested">The date the request was made as a <see cref="DateTimeOffset"/>.</param>
    /// <returns>The <see cref="SigningContentBuilder"/> instance.</returns>
    internal SigningContentBuilder WithDateRequested(DateTimeOffset dateRequested)
    {
        Context.DateRequested = dateRequested;
        return this;
    }

    /// <summary>
    /// Sets the nonce value to be used in the signing content.
    /// </summary>
    /// <param name="nonce">The nonce as a <see cref="Guid"/>.</param>
    /// <returns>The <see cref="SigningContentBuilder"/> instance.</returns>
    internal SigningContentBuilder WithNonce(Guid nonce)
    {
        Context.Nonce = nonce;
        return this;
    }

    /// <summary>
    /// Sets the header values to be included in the signing content.
    /// </summary>
    /// <param name="headerValues">The header values as a collection of <see cref="HeaderValue"/>.</param>
    /// <returns>The <see cref="SigningContentBuilder"/> instance.</returns>
    internal SigningContentBuilder WithHeaderValues(IReadOnlyCollection<HeaderValue> headerValues)
    {
        ArgumentNullException.ThrowIfNull(headerValues, nameof(headerValues));
        
        Context.HeaderValues = headerValues;
        return this;
    }

    /// <summary>
    /// Sets the content hash to be used in the signing content.
    /// </summary>
    /// <param name="contentHash">The content hash as a string.</param>
    /// <returns>The <see cref="SigningContentBuilder"/> instance.</returns>
    internal SigningContentBuilder WithContentHash(string? contentHash)
    {
        Context.ContentHash = contentHash;
        return this;
    }

    /// <summary>
    /// Creates a new instance of the <see cref="SigningContentBuilder"/>.
    /// </summary>
    /// <returns>A new <see cref="SigningContentBuilder"/> instance.</returns>
    public virtual SigningContentBuilder CreateBuilder() => new SigningContentBuilder();

    /// <summary>
    /// Builds and returns the signing content as a string.
    /// </summary>
    /// <returns>The signing content as a string.</returns>
    public virtual string Build()
    {
        ArgumentNullException.ThrowIfNull(Context.Request, nameof(Context.Request));

        Builder.Append($"{Context.Request.Method}");

        if (Context.Request.RequestUri is not null)
        {
            // The problem above can be replicated when signing with HmacManager directly,
            // then proceeding to call Send/SendAsync on an HttpClient. When signing manually, without
            // using the HmacDelegatingHandler, if a relative URI is specified on the HttpRequestMessage,
            // then this builder only has access to that relative URI. However, on the recieving end,
            // the authentication handler will rebuild the signing content with the absolute URI because
            // it has been sent with the BaseAddress specified on the HttpClient. This causes a 401 response
            // because the signatures won't match.

            // Method #1 - Fallback to looking at the host header, if that header isn't present,
            // throw an exception forcing users to manually add the host.

            // Method #2 - Throw an exception immediately if the RequestUri.IsAbsoluteUri is false

            // This can be configurable during Program.cs, something like RequireAbsoluteUri().
            if (Context.Request.RequestUri.IsAbsoluteUri)
            {
                Builder.Append($":{Context.Request.RequestUri.PathAndQuery}");

                // Authority is not available in a relative URI, so we remove it from the signing content for consistency.
                //Builder.Append($":{Context.Request.RequestUri.Authority}");
            }
            else
            {
                Builder.Append($":{Context.Request.RequestUri.OriginalString}");
            }
        }

        Builder.Append($":{Context.DateRequested?.ToUnixTimeMilliseconds()}");
        Builder.Append($":{Context.PublicKey}");

        if (Context.ContentHash is not null)
        {
            Builder.Append($":{Context.ContentHash}");
        }

        if (Context.HeaderValues.Any())
        {
            Builder.Append(":");
            Builder.AppendJoin(":", Context.HeaderValues.Select(element => element.Value));
        }

        Builder.Append($":{Context.Nonce}");
        return Builder.ToString();
    }
}

File: client/lib/src/builders/signing-content-builder.ts
import SigningContentContext from "./signing-content-context";

/**
 * Builder for constructing the signing content used in HMAC authentication.
 */
export default class SigningContentBuilder {
    protected context: SigningContentContext = {
        request: null,
        contentHash: null,
        dateRequested: new Date(),
        nonce: "",
        publicKey: "",
        signedHeaders: []
    };

    withRequest(request: Request) {
        this.context.request = request;
        return this;
    }

    /** 
     * Adds the public key to the signing content.
     * @param publicKey The public key to include.
     * @returns The instance of the builder for method chaining.
     */
    withPublicKey = (publicKey: string) => {
        this.context.publicKey = publicKey;
        return this;
    }

    /** 
     * Adds the date when the request was made to the signing content.
     * @param dateRequested The date of the request.
     * @returns The instance of the builder for method chaining.
     */
    withDateRequested = (dateRequested: Date) => {
        this.context.dateRequested = dateRequested;
        return this;
    }

    /** 
     * Adds the content hash to the signing content.
     * @param contentHash The content hash to include.
     * @returns The instance of the builder for method chaining.
     */
    withContentHash = (contentHash: string | null) => {
        this.context.contentHash = contentHash;
        return this;
    }

    /** 
     * Adds a nonce (unique identifier) to the signing content.
     * @param nonce The nonce to include.
     * @returns The instance of the builder for method chaining.
     */
    withNonce = (nonce: string) => {
        this.context.nonce = nonce;
        return this;
    }

    /** 
     * Adds signed headers to the signing content.
     * Throws an error if any required headers are missing.
     * @param signedHeaders The headers that are signed.
     * @param headers The Headers object to extract header values from.
     * @returns The instance of the builder for method chaining.
     */
    withSignedHeaders = (signedHeaders: string[] = []) => {
        if (signedHeaders.length) {
            const headerValues = signedHeaders.map(key => this.context.request?.headers.get(key));
            const isMissingHeaderValues = headerValues.some(value => value === null);
            if (isMissingHeaderValues) {
                throw new Error("There are missing headers that are required for signing.");
            } else {
                this.context.signedHeaders = headerValues as string[];
            }
        }

        return this;
    }

    createBuilder = () => new SigningContentBuilder();

    /** 
     * Builds the signing content by joining the individual components.
     * @returns The constructed signing content as a string.
     */
    build = async () => {
        const { method, url } = this.context.request as Request;
        const { host, search, pathname } = new URL(url);

        const stringBuilder: string[] = [];
        stringBuilder.push(method);
        stringBuilder.push(`${pathname}${search}`);
        stringBuilder.push(host);
        stringBuilder.push(this.context.dateRequested.getTime().toString());
        stringBuilder.push(this.context.publicKey);
        
        if (this.context.contentHash) {
            stringBuilder.push(this.context.contentHash);
        }

        stringBuilder.push(...this.context.signedHeaders);
        stringBuilder.push(this.context.nonce);
        return stringBuilder.join(":");
    }
}